src/LabelGroup/LabelGroup.tsx

```diff
@@ -1,17 +1,365 @@
+
+          for (const entry of entries) {
+            // Checks which children are intersecting the root container
+            const targetId = entry.target.getAttribute('data-index')
+            if (targetId) {
+              updatedEntries[targetId] = entry.isIntersecting
+            }
+          }
+
+          // Updates the visibility map based on the intersection results.
+          setVisibilityMap(prev => ({
+            ...prev,
+            ...updatedEntries,
+          }))
+        },
+        {
+          root: containerRef.current,
+          rootMargin: `0px -${buttonClientRect.width}px 0px 0px`,
+          threshold: 1,
+        },
+      )
+
+      for (const item of containerRef.current?.children || []) {
+        if (item.getAttribute('data-index')) {
+          observer.observe(item)
+        }
+      }
+
+      return () => observer.disconnect()
+    }
+    // We're not auto truncating, so we need to hide children after the given `visibleChildCount`.
+    else {
+      hideChildrenAfterIndex(visibleChildCount)
+    }
+  }, [buttonClientRect, visibleChildCount, hideChildrenAfterIndex])
+
+  // Updates the index of the first hidden child.
+  // We need to keep track of this so we can focus the first hidden child when the overflow is shown inline.
+  React.useEffect(() => {
+    // If we're using an overlay, we don't need to keep track of the first hidden index.
+    if (overflowStyle === 'overlay') {
+      return
+    }
+    if (hiddenItemIds.length) {
+      firstHiddenIndexRef.current = parseInt(hiddenItemIds[0], 10)
+    }
+  }, [hiddenItemIds, overflowStyle])
+
+  // Updates the index of the first hidden child.
+  // We need to keep track of this so we can focus the first hidden child when the overflow is shown inline.
+  React.useEffect(() => {
+    if (overflowStyle === 'overlay') {
+      return
+    }
+    const firstHiddenChildDOM = document.querySelector<HTMLElement>(`[data-index="${firstHiddenIndexRef.current}"]`)
+    const focusableChild = firstHiddenChildDOM ? getFocusableChild(firstHiddenChildDOM) : null
+
+    if (isOverflowShown) {
+      // If the first hidden child is focusable, focus it.
+      // Otherwise, focus the collapse button.
+      if (focusableChild) {
+        focusableChild.focus()
+      } else {
+        collapseButtonRef.current?.focus()
+      }
+    }
+  }, [overflowStyle, isOverflowShown])
+
+  // If truncation is enabled, we need to render based on truncation logic.
+  return visibleChildCount ? (
+    <StyledLabelGroupContainer
+      ref={containerRef}
+      data-overflow={overflowStyle === 'inline' && isOverflowShown ? 'inline' : undefined}
+      sx={sxProp}
+    >
+      {React.Children.map(children, (child, index) => {
+        if (hiddenItemIds.includes(index.toString())) {
+          // hidden children
+          return (
+            <div
+              // data-index is used as an identifier we can use in the IntersectionObserver
+              data-index={index}
```

Comment:
I wonder if using index as an identifier could create bugs if the labels change order (example: if tokens can be dragged and dropped).